dan_layer/core/src/workers/hotstuff_waiter.rs

```diff
@@ -653,12 +666,72 @@ where
         }
     }
 
-    fn validate_qc(&self, _qc: &QuorumCertificate) -> Result<(), HotStuffError> {
-        // TODO: get committee at epoch
-        // TODO: Validate committee signatures
+    async fn validate_qc(&self, qc: &QuorumCertificate) -> Result<(), HotStuffError> {
+        // extract all the pairs of signer-signature present in the QC
+        let signer_signatures = Self::extract_signer_signatures_from_qc(qc)?;
+
+        // the QC should not have repeated signers
+        let signers_iter = signer_signatures.iter().map(|s| NodeAddressable::as_bytes(&s.0));
+        let signers_set: HashSet<&[u8]> = HashSet::from_iter(signers_iter);
+        if signer_signatures.len() != signers_set.len() {
+            return Err(HotStuffError::InvalidQuorumCertificate(
+                "duplicated signers".to_string(),
+            ));
+        }
+
+        // check that the minimum quorum has been reached
+        let committee = self.epoch_manager.get_committee(qc.epoch(), qc.shard()).await?;
+        if signers_set.len() < committee.consensus_threshold() {
+            return Err(HotStuffError::InvalidQuorumCertificate(
+                "insufficient quorum".to_string(),
+            ));
+        }
+
+        // All signers must be included in the epoch commitee for the shard
+        let commitee_iter = committee.members.iter().map(|m| m.as_bytes());
+        let commitee_set: HashSet<&[u8]> = HashSet::from_iter(commitee_iter);
```

Comment:
I think you can pass directly `committee.members.iter().map(|m| m.as_bytes()).collect::<HashSet<_>>()`.